app/src/main/java/com/eis/geoCalendar/database/EventDatabasePool.java

```diff
@@ -0,0 +1,257 @@
+ * This means the databases for Event type A will be separated from Event type B.
+ * To avoid losing the Object Pool behaviour when extending, the class has been made {@code final}.
+ *
+ * @param <E> The type of event this Database stores.
+ * @author Riccardo De Zen
+ * @author Giorgia Bortoletti
+ */
+
+public final class EventDatabasePool<E extends Event> implements EventDatabase<E> {
+
+    /**
+     * The following String ensures a combination of {@link TypeToken} and database name won't
+     * collide. Because it cannot be part of a {@link Class}' name.
+     */
+    private static final String SEPARATOR = "-";
+    //Prefix used to reduce chances of collisions with any other class' Room Databases.
+    private static final String DB_NAME_PREFIX = "geoCalendar-database";
+    //String to be formatted with the appropriate parameters.
+    private static final String DB_NAME_TEMPLATE =
+            DB_NAME_PREFIX + SEPARATOR + "%1$s" + SEPARATOR + "%2$s";
+
+    //Map containing all the active instances for this Database class.
+    private static Map<String, EventDatabasePool> activeInstances = new ArrayMap<>();
+
+    private AbstractEventDatabase physicalDatabase;
+    private TypeToken<E> eventType;
+    private JsonEventParser<E> parser;
+    private String name;
+
+    /**
+     * Only available constructor.
+     *
+     * @param eventType        A reference to the Class this Database stores.
+     * @param name             The name for this Database.
+     * @param physicalDatabase The actual {@link AbstractEventDatabase} this Object should access.
+     */
+    private EventDatabasePool(TypeToken<E> eventType, String name,
+                              AbstractEventDatabase physicalDatabase) {
+        this.physicalDatabase = physicalDatabase;
+        this.eventType = eventType;
+        this.name = name;
+        this.parser = new JsonEventParser<>(eventType);
+    }
+
+    /**
+     * Method to retrieve an instance of the Database.
+     * The suppressed warning refers to the cast in the return statement. Such a cast can be
+     * considered safe (as of now) because this class ensures the instance with a certain key
+     * actually contains the appropriate Event type.
+     *
+     * @param context   The calling Context, needed to instantiate the Database.
+     * @param eventType The stored {@code Event} class' {@link TypeToken}.
+     * @param name      The name for the Database.
+     * @param <E>       The stored {@code Event} type.
+     * @return The existing appropriate instance, if possible, or a newly created appropriate
+     * instance.
+     * @see EventDatabasePool#generateName(TypeToken, String) For more info on why the cast is
+     * safe.
+     */
+    @SuppressWarnings("unchecked")
+    public static <E extends Event> EventDatabase<E> getInstance(Context context,
+                                                                 TypeToken<E> eventType,
```

Comment:
Neat use of `TypeToken`

Reply:
I still don't have half a clue what I'm doing tho